### ISSUE TITLE
Question about implementing client-level DP

### ISSUE THREAD

**Issue by anirban-nath** [wenzhu23333/Differential-Privacy-Based-Federated-Learning#9]
I am trying to implement client-level differential privacy where noise is added during the federated averaging step. To estimate epsilon at the step, I am using say gradient clipping `S` and a noise multiplier `z`. From there I calculate `sigma = z*S/q`, where `q` is the sampling rate of clients (if um_clients = 5 and q = 0.6, then the clients participating in federation each round = 5 * 0.6 = 3). 

So now, I am using this sigma parameter as variance for adding Gaussian noise, like `N(0, sigma*sigma)`. I am using tensorflow privacy to estimate `epsilon` value, where it takes as input `(sampling rate, noise multiplier, epochs, and delta=1e-5).` What should I set as the value of `epochs` for client-level DP? Number of federation rounds (maybe 10) or the number of local epochs of all clients (maybe 300 per client)?

**Comment by wenzhu23333** [comment]
In the DP-SGD algorithm, epoch refers to the number of sampling lot. I think the sampling lot in your algorithm is the sampling client, so it should be Number of federation rounds. But I'm not sure what your specific noise mechanism is? Are you adding noise client-side or server-side?
Refer to this paper: K. Wei et al., "User-Level Privacy-Preserving Federated Learning: Analysis and Performance Optimization," in IEEE Transactions on Mobile Computing, vol. 21, no. 9, pp. 3388-3401, 1 Sept. 2022, doi: 10.1109/TMC.2021.3056991.

**Comment by anirban-nath** [comment]
>  But I'm not sure what your specific noise mechanism is? Are you adding noise client-side or server-side?
Hi. My noise mechanism is server-side noise for client-level DP. This is why I am adding noise during the federation procedure.

**Comment by wenzhu23333** [comment]
That should be **Number of federation rounds**.
